instaclustr/resource_kafka_topic.go

```diff
@@ -0,0 +1,398 @@
+// There are two use cases of resourceKafkaTopicUpdate:
+// 1. Providing all the possible configs in resource, and modify the ones you want to update. Then keeping all the
+// configs in the resource.
+// 2. Providing some configs (not all) you want to update, and then keep only these configs in the resource.
+// Terraform always detects that there is an update in case 2 if you run the plan again, because resourceKafkaTopicRead
+// reads all the configs, but there are only some of them provided in resource. For the missing configs, Terraform just
+// sets them as their default values.
+//
+// Although Terraform detects an update, func getChangedConfigMap is used to handle case 2. It makes sure that there
+// will not be a real update if the configs in the resources are not changed. So, it's safe to keep only partial of
+// configs in the config{} block, there won't be an update request if their values are same as what have been read.
+func resourceKafkaTopicUpdate(d *schema.ResourceData, meta interface{}) error {
+	cluster_id := d.Get("cluster_id").(string)
+	topic := d.Get("topic").(string)
+	log.Printf("[INFO] Updating the config of Kafka topic %s in %s.", topic, cluster_id)
+
+	// This is for when only replication_factor or partitions are changed, but we don't support changing them currently
+	if !d.HasChange("config") {
+		log.Printf("[INFO] Currently we only support updating topic's config. There are no changes in topic %s's config.", topic)
+		return nil
+	}
+	// This is for when there is no config{} block in the resource, or the whole config{} is removed from the resource.
+	if len(d.Get("config").([]interface{})) == 0 {
+		log.Printf("[INFO] There are no configs to be updated.")
+		return nil
+	}
+	client := meta.(*Config).Client
+
+	changedConfigMap, err := getChangedConfigMap(d)
+	if err != nil {
+		return fmt.Errorf("[Error] Error getting the changed config map for topic %s: %w", topic, err)
+	}
+	if len(changedConfigMap) == 0 {
+		log.Printf("[INFO] There are no configs to be updated.")
+		err = resourceKafkaTopicRead(d, meta)
+		if err != nil {
+			return fmt.Errorf("[Error] Error reading the updated topic config for topic %s: %w", topic, err)
+		}
+		return nil
+	}
+
+	var kafkaTopicConfigOptions KafkaTopicConfigOptions
+	err = mapstructure.Decode(changedConfigMap, &kafkaTopicConfigOptions)
+	if err != nil {
+		return fmt.Errorf("[Error] Error decoding the changed config map to KafkaTopicConfigOptions for topic %s: %w", topic, err)
+	}
+
+	updateKafkaTopicRequest := UpdateKafkaTopicRequest{
+		Config: &kafkaTopicConfigOptions,
+	}
+
+	jsonStr, err := json.Marshal(updateKafkaTopicRequest)
+	if err != nil {
+		return fmt.Errorf("[Error] Error creating kafka topic update request: %w", err)
+	}
+
+	err = client.UpdateKafkaTopic(cluster_id, topic, jsonStr)
+	if err != nil {
+		return fmt.Errorf("[Error] Error updating the config of Kafka topic %s: %w", topic, err)
+	}
+
+	//Reading updated configs
+	err = resourceKafkaTopicRead(d, meta)
```

Comment:
Why Is there a need to manually invoke the reading life cycle?

Reply:
This is basically to keep the terraform.tfstate in sync. Especially when there are only partial configs are provided in the resource. This read after update will not set the missing configs as their zero-values, instead it reads the real values of them.